File: 03_Two_Sum.ts
// https://neetcode.io/problems/two-integer-sum
class Solution_03 {
  /**
   * @param {number[]} nums
   * @param {number} target
   * @return {number[]}
   */
  twoSum(nums: number[], target: number) {
    const hashMap = {};
    for (let i=0; i<nums.length; i++) {
      let diff: number = target - nums[i];
      if (hashMap.hasOwnProperty(diff)) {
        return [hashMap[diff], i];
      }
      hashMap[nums[i]] = i;
    }
  }
}

File: Arrays and Hashmaps/03_Two_Sum.cs
public class Solution {
    // Brute Force solution
    // public int[] TwoSum(int[] nums, int target) {
    //     for (int i=0; i < nums.Length; i++) {
    //         for (int j=i+1; j < nums.Length; j++) {
    //             if (nums[i] + nums[j] == target) {
    //                 return new int[] {i, j};
    //             }
    //         }
    //     }
    //     return new int[0];
    // }

    // Hash Map solution
    public int[] TwoSum(int[] nums, int target) {
        Dictionary<int, int> hashMap = new Dictionary<int, int>();

        for(int i=0; i<nums.Length; i++) {
            int diff = target - nums[i];

            if (hashMap.ContainsKey(diff)) {
                return new int[]{hashMap[diff], i};
            } else {
                hashMap.Add(nums[i], i);
            }
        }
        return new int[0];
    }
}